azurelinuxagent/ga/cgroupapi.py

```diff
@@ -103,28 +109,78 @@ def _foreach_legacy_cgroup(operation):
 
                 if os.path.exists(procs_file):
                     procs_file_contents = fileutil.read_file(procs_file).strip()
-                    daemon_pid = CGroupsApi.get_daemon_pid()
+                    daemon_pid = CGroupUtil.get_daemon_pid()
 
                     if ustr(daemon_pid) in procs_file_contents:
                         operation(controller, daemon_pid)
         finally:
             for _, cgroup in legacy_cgroups:
-                logger.info('Removing {0}', cgroup)
+                log_cgroup_info('Removing {0}'.format(cgroup), send_event=False)
                 shutil.rmtree(cgroup, ignore_errors=True)
         return len(legacy_cgroups)
 
     @staticmethod
-    def get_daemon_pid():
-        return int(fileutil.read_file(get_agent_pid_file_path()).strip())
+    def cleanup_legacy_cgroups():
+        """
+        Previous versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent;
+        starting from version 2.2.41 we track the agent service in walinuxagent.service instead of WALinuxAgent/WALinuxAgent. If
+        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
+        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        """
+        return CGroupUtil._foreach_legacy_cgroup(lambda *_: None)
 
 
-class SystemdCgroupsApi(CGroupsApi):
+class SystemdRunError(CGroupsException):
     """
-    Cgroups interface via systemd
+    Raised when systemd-run fails
     """
 
+    def __init__(self, msg=None):
+        super(SystemdRunError, self).__init__(msg)
+
+
+def get_cgroup_api():
+    """
+    Determines which version of Cgroup should be used for resource enforcement and monitoring by the Agent and returns
+    the corresponding Api.
+
+    Uses 'stat -f --format=%T /sys/fs/cgroups' to get the cgroup hierarchy in use.
+        If the result is 'cgroup2fs', cgroup v2 is being used.
+        If the result is 'tmpfs', cgroup v1 or a hybrid mode is being used.
+            If the result of 'stat -f --format=%T /sys/fs/cgroup/unified' is 'cgroup2fs', then hybrid mode is being used.
+
+    Raises exception if an unknown mode is detected. Also raises exception if hybrid mode is detected and there are
+    controllers available to be enabled in the unified hierarchy (the agent does not support cgroups if there are
+    controllers simultaneously attached to v1 and v2 hierarchies).
+    """
+    root_hierarchy_mode = shellutil.run_command(["stat", "-f", "--format=%T", CGROUP_FILE_SYSTEM_ROOT]).rstrip()
```

Comment:
we assumed the root file system will be /sys/fs/cgroup. But one can mount the cgroup filesystem anywhere right, so according to this check if its not found in expected path we don't use cgroups? will this check fails if path not found?

Reply:
The other piece that is contradicting is we don't assume root path there and finding root path with mount command in each cgroups api implementation.

Reply:
I'll be adding a check that the path exists. If not, we won't use cgroups since we expect systemd to mount cgroup filesystem there. We'll collect telemetry so we can get an idea for how many VMs are mounting the cgroup filesystem elsewhere.